STSampling.C

```diff
@@ -0,0 +1,565 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2019, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+#include "Matrix.h"
+#include "mpi.h"
+#include <cmath>
+#include <map>
+#include <set>
+
+#include "STSampling.h"
+#include "DEIM.h"
+
+using namespace std;
+
+namespace CAROM {
+
+// This function implements Algorithm 2 of Choi and Carlberg 2019.
+// All spatial indices are used here.
+// Spatial basis s_basis is distributed, of size (number of spatial DOFs) x (number of basis vectors)
+// Temporal basis t_basis is stored in its entirety on each process, of size (number of temporal DOFs) x (number of basis vectors)
+// In general, there may be multiple temporal basis vectors associated with each spatial basis vector, but here we assume for now
+// that there is only one, i.e. column j of s_basis corresponds to column j of t_basis.
+void SampleTemporalIndices(const Matrix* s_basis,
+                           const Matrix* t_basis,
+                           const int num_f_basis_vectors_used,
+                           int* t_samples,
+                           const int myid,
+                           const int num_procs,
+                           const int num_samples_req,
+                           const bool excludeFinalTime)
+{
+    CAROM_VERIFY(t_basis->distributed());
+
+    // Get the number of basis vectors and the size of each basis vector.
+    CAROM_VERIFY(0 < num_f_basis_vectors_used && num_f_basis_vectors_used <= s_basis->numColumns()
+                 && num_f_basis_vectors_used <= t_basis->numColumns());
+    CAROM_VERIFY(num_samples_req > 0);  // TODO: just replace num_samples with num_samples_req
+    const int num_basis_vectors =
+        std::min(num_f_basis_vectors_used, s_basis->numColumns());
+    //const int num_samples = num_samples_req > 0 ? num_samples_req : num_basis_vectors;
+    const int num_samples = num_samples_req;
+    const int numExcluded = excludeFinalTime ? 1 : 0;
+    //CAROM_VERIFY(num_basis_vectors <= num_samples && num_samples <= t_basis->numRows() - numExcluded);
+    CAROM_VERIFY(num_samples <= t_basis->numRows() - numExcluded);
+    const int s_size = s_basis->numRows();
+    const int t_size = t_basis->numRows();
+
+    const int ns_mod_nr = num_samples % num_basis_vectors;
+    int ns = 0;
+
+    // The small matrix inverted by the algorithm.  We'll allocate the largest
+    // matrix we'll need and set its size at each step in the algorithm.
+    Matrix M(num_basis_vectors, num_basis_vectors, false);  // TODO: is this big enough to avoid reallocations?
```

Comment:
You mentioned below that M should be declared as distributed. Should we do that here, or is this PR only for serial?

Reply:
This PR is only for the serial case. There will be a significant amount of work to get the parallel case, so I am just trying to document some of the things to do.